### PR TITLE
[SYCL] Add Level-Zero interop with specification of ownership for Queue.

### PR DESCRIPTION
**Pull request by rbegam**
The L0 backend interop functions take the ownership of the passed L0 handles. This is problematic for some cases as the users can not use the original handle when they go out of scope and destroyed.
This PR provides a way for users of SYCL - L0 interop to retain the ownership of the L0 handles for the Queue. 

Signed-off-by: rehana begam <rehana.begam@intel.com>

### CODE REVIEW

File: sycl/include/CL/sycl/detail/pi.h [hunk 1]
```diff
@@ -36,10 +36,11 @@
 // 2. A number of types needed to define pi_device_binary_property_set added.
 // 3. Added new ownership argument to piextContextCreateWithNativeHandle.
 // 4. Add interoperability interfaces for kernel.
+// 5. Added new ownership argument to piextQueueCreateWithNativeHandle.
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h:39]: The numbering (5.) is not in sync with PI versions

File: sycl/include/CL/sycl/backend.hpp [hunk 2]
```diff
@@ -91,7 +91,7 @@ __SYCL_EXPORT context make_context(pi_native_handle NativeHandle,
                                    const async_handler &Handler,
                                    backend Backend);
 __SYCL_EXPORT queue make_queue(pi_native_handle NativeHandle,
-                               const context &TargetContext,
+                               const context &TargetContext, bool KeepOwnership,
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/backend.hpp:94]: I think we shouldn't extend this one (shared between different backends) as we only extend level_zero. Also I see there is no "keep-ownership" for make_context just above.

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 3]
```diff
@@ -2149,6 +2149,7 @@ pi_result cuda_piextQueueGetNativeHandle(pi_queue queue,
 ///
 /// \return TBD
 pi_result cuda_piextQueueCreateWithNativeHandle(pi_native_handle, pi_context,
+                                                bool ownNativeHandle,
```
Comment (smaslov-intel) [inline sycl/plugins/cuda/pi_cuda.cpp:2152]: you should probably add a fake use like `(void)ownNativeHandle'`

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 4]
```diff
@@ -2471,7 +2471,7 @@ pi_result piQueueRelease(pi_queue Queue) {
     std::lock_guard<std::mutex> Lock(Queue->PiQueueMutex);
     Queue->RefCount--;
     if (Queue->RefCount == 0)
-      RefCountZero = true;
+      RefCountZero = Queue->OwnZeCommandQueue ? true : false;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:2474]: This way you will not execute the entire if below, whereas you only need not run the `zeCommandQueueDestroy` at the end

File: sycl/include/CL/sycl/backend.hpp (hunk 2 above)
Comment (smaslov-intel) [reply sycl/include/CL/sycl/backend.hpp:94]: Actually, I am fine with this change (after seeing that level_zero::make_queue just redirects to this detail::make_queue)

**Review by smaslov-intel — CHANGES_REQUESTED** [review]

File: sycl/include/CL/sycl/detail/pi.h (hunk 1 above)
Comment (rbegam) [reply sycl/include/CL/sycl/detail/pi.h:39]: should it match with the major/minor version? I've followed the inst from line#25 above and increased both by 1.

Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.h:39]: Following instructions in line #25 is all fine. Just let's have the comments about the changes tell what was the version where it was made. So, in your case I'd say have "5" changed to "4.6". And, please add a comment about this numbering in the comment above.

File: sycl/include/CL/sycl/backend.hpp (hunk 2 above)
Comment (alexbatashev) [reply sycl/include/CL/sycl/backend.hpp:94]: @smaslov-intel is it ok that this change breaks ABI? Shouldn't we keep the old version of this function as well?

Comment (smaslov-intel) [reply sycl/include/CL/sycl/backend.hpp:94]: Yes, let's keep the old version (to be removed at major release 2022.0) and link that to the new one (with keep-ownership = false).

File: sycl/plugins/cuda/pi_cuda.cpp (hunk 3 above)
Comment (rbegam) [reply sycl/plugins/cuda/pi_cuda.cpp]: done

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 4 above)
Comment (rbegam) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: done

File: sycl/include/CL/sycl/detail/pi.h (hunk 1 above)
Comment (rbegam) [reply sycl/include/CL/sycl/detail/pi.h]: done.

**Comment by rbegam** [comment]
ping @smaslov-intel

File: sycl/include/CL/sycl/backend.hpp (hunk 2 above)
Comment (smaslov-intel) [reply sycl/include/CL/sycl/backend.hpp:94]: I don't see that you keep a version without the new "bool" argument

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 5]
```diff
@@ -2494,11 +2494,14 @@ pi_result piQueueRelease(pi_queue Queue) {
...
       Queue->ZeCommandListFenceMap.clear();
-      ZE_CALL(zeCommandQueueDestroy, (Queue->ZeComputeCommandQueue));
-      Queue->ZeComputeCommandQueue = nullptr;
-      if (Queue->ZeCopyCommandQueue) {
-        ZE_CALL(zeCommandQueueDestroy, (Queue->ZeCopyCommandQueue));
-        Queue->ZeCopyCommandQueue = nullptr;
+
+      if (Queue->OwnZeCommandQueue) {
+        ZE_CALL(zeCommandQueueDestroy, (Queue->ZeComputeCommandQueue));
+        Queue->ZeComputeCommandQueue = nullptr;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:2500]: I think we should still null-ify these, even if we don't own the handles

**Review by smaslov-intel — CHANGES_REQUESTED** [review]
few more comments

File: sycl/include/CL/sycl/backend.hpp (hunk 2 above)
Comment (rbegam) [reply sycl/include/CL/sycl/backend.hpp:94]: right, adding it.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 5 above)
Comment (rbegam) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: done.

File: sycl/include/CL/sycl/backend.hpp (hunk 2 above)
Comment (rbegam) [reply sycl/include/CL/sycl/backend.hpp:94]: done.

**Comment by rbegam** [comment]
hi @smaslov-intel, could you please review? the pre-commit fail is for a time-out and I've already restarted it. thanks.

**Review by smaslov-intel — APPROVED** [review]
LGTM

**Comment by rbegam** [comment]
I re-ran the precommit tests and there are no time outs. all tests are passing. 
http://icl-jenkins.sc.intel.com:8080/blue/organizations/jenkins/SYCL_CI%2Fintel%2FWin%2FTest_Suite/detail/Test_Suite/5289/pipeline/

**Review by jbrodman — APPROVED** [review]

**Pull request merged**